rpc/lib/server/ws_handler.go

```diff
@@ -0,0 +1,423 @@
+package rpcserver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"runtime/debug"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/pkg/errors"
+
+	amino "github.com/tendermint/go-amino"
+
+	cmn "github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/libs/log"
+	types "github.com/tendermint/tendermint/rpc/lib/types"
+)
+
+///////////////////////////////////////////////////////////////////////////////
+// WebSocket handler
+///////////////////////////////////////////////////////////////////////////////
+
+// WebsocketManager provides a WS handler for incoming connections and passes a
+// map of functions along with any additional params to new connections.
+// NOTE: The websocket path is defined externally, e.g. in node/node.go
+type WebsocketManager struct {
+	websocket.Upgrader
+
+	funcMap       map[string]*RPCFunc
+	cdc           *amino.Codec
+	logger        log.Logger
+	wsConnOptions []func(*wsConnection)
+}
+
+// NewWebsocketManager returns a new WebsocketManager that passes a map of
+// functions, connection options and logger to new WS connections.
+func NewWebsocketManager(funcMap map[string]*RPCFunc, cdc *amino.Codec, wsConnOptions ...func(*wsConnection)) *WebsocketManager {
+	return &WebsocketManager{
+		funcMap: funcMap,
+		cdc:     cdc,
+		Upgrader: websocket.Upgrader{
+			CheckOrigin: func(r *http.Request) bool {
+				// TODO ???
```

Comment:
???

Reply:
The [default behaviour](https://github.com/gorilla/websocket/blob/483fb8d7c32fcb4b5636cd293a92e3935932e2f4/server.go#L87) would be relevant to browser-based clients, afaik. I suppose having a pass-through is a workaround for allowing for more complex security schemes, shifting the burden of AuthN/AuthZ outside the Tendermint RPC.

I can't think of other uses right now that would warrant a `TODO` though. The real backstory of this `TODO` shall remain shrouded in mystery 👻

Reply:
Gold! Saved this comment for future generations 😄